### PR TITLE
Async Request Class

### PR DESCRIPTION
**Pull request by hibestil**
# Description
Taking inspiration from [the discussion](https://github.com/gradio-app/gradio/pull/1022/files#r870856743) on https://github.com/gradio-app/gradio/pull/1022, I implemented a new Request class. 

The proposed Request class is a low-level API that allows you to create asynchronous HTTP requests without a context manager. Compared to making calls by using `httpx.Request` directly, the Request includes `httpx.Request`’s all functionalities plus it offers more flexibility and control over:

- Includes response validation functionality both using validation models and functions.
- Exceptions are handled silently during the request call, allowing us to inspect each one individually in the case of multiple asynchronous request calls and some of them failing. ([with inspiration from aiohttp's hdrs methods](https://github.com/gradio-app/gradio/pull/1022#discussion_r870831233))
- Provides HTTP request types with `Request.Method` (an enum class) for ease of usage

The Request also offers some util functions such as _has_exception_, _is_valid_, and _status_ to inspect and get detailed information about executed request calls.

**The basic usage of  the Request** is as follows: 
```python
from gradio.utils import Request

# Make a request call by creating a Request object with inputs (method, URL, etc.), and get the response
client_response = await Request(method=Request.Method.GET, url="http://headers.jsontest.com/")
# And then you can use util functions to do some post request checks depending on your case. (here we used .isvalid())
if client_response.is_valid():
     # Finally, call the validated_data property to get the response data.
     response = client_response.validated_data
     # do some operations with the response
```

The changes so far only include the Request class and its basic level tests. Wanted to get your opinion on this before changing the other files. What are your thoughts?

[@FarukOzderim](https://github.com/FarukOzderim) [@aliabid94](https://github.com/aliabid94) 

### Additional Dependencies:
- pydantic (to validate responses with pydantic models)
- httpx ([see the reason](https://github.com/gradio-app/gradio/pull/1022#issuecomment-1102755297))

### Closes:
-  [Gradio Async Support #727](https://github.com/gradio-app/gradio/issues/727)

# Checklist:

- [x] I have performed a self-review of my own code
- [x] My code follows the style guidelines of this project
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: test/requirements.in [hunk 1]
```diff
@@ -18,3 +18,5 @@ black
 isort
 flake8
 torch
+httpx
```
Comment (FarukOzderim) [inline test/requirements.in:21]: This is just test dependencies, if a lib is used within the Gradio library, you also need to add them to the setup.py and requires.txt

**Review by FarukOzderim — CHANGES_REQUESTED** [review]
I think this looks very decent overall, thx for the contribution @hibestil ❤️



Would like to take thoughts from @aliabid94 and @abidlabs about the wrapper, and finalize this PR.

**Comment by FarukOzderim** [comment]
>if client_response.is_valid():
     # Finally, call the validated_data property to get the response data.
     response = client_response.validated_data
     # do some operations with the response


Also I would prefer to have this in one line if possible. Can we have smt like this?
# Returns None or some specificied error
response = client_response.get_validated_data()

**Comment by FarukOzderim** [comment]
@hibestil  Tests are failing, can you run `bash scripts/format_backend` and fix those cases?

File: test/requirements.in (hunk 1 above)
Comment (hibestil) [reply test/requirements.in:21]: Oh, I forgot that. Now, it's fixed 👍

**Comment by hibestil** [comment]
> > if client_response.is_valid():
> > # Finally, call the validated_data property to get the response data.
> > response = client_response.validated_data
> > # do some operations with the response
> 
> Also I would prefer to have this in one line if possible. Can we have smt like this?
> 
> ```
> # Returns None or some specificied error
> response = client_response.get_validated_data()
> ```

If we create a function named get_validated_data and if it returns validated data or the exception. How will we behave for the return value of the function? Here is an example:
```python
class Request:
   ...
   def get_validated_data() -> Union[Json,Exception]:
       try:
              return validate_by_model()
       except:
              return Exception 
       return validated_data if the data is valid, else return the Exception instance
   ...

# Case 1 : Response is valid.
response = await Request(...)
# Let's assume response is valid
validated_data = response.validated_data
user_name = validated_data["user_name"]

# Case 2: Response is not valid. 
# In this case we should handle/check exceptions
response = await Request(...)
# Let's assume response is not valid
validated_data = response.validated_data # Returns the Exception
# user_name = validated_data["user_name"] # It will raise an error. So we should check if there is an exception:
if isinstance(validated_data, Exception):
   # handle the Exception
else:
    user_name = validated_data["user_name"] 

```

BUT, In the current implementation, if a problem occurs while making a request, the exception is stored in the Request._exception, and then you get the validated data as None. So we can handle this in one line, Using the is_valid function is completely up to you.
```python


client_response = await Request(method=Request.Method.GET, url="http://headers.jsontest.com/")
# Let's assume that an exception is thrown in the request call
response = client_response.validated_data #  response will be None

# You don't have to use is_valid function.
# But if you want to raise an exception for the problem then you can 
if not response.is_valid():
    raise response.exception
```

So I think we should still have both get_validated_data (validated_data) and is_valid functions to easily manage and handle exceptions. But I liked creating a method (get_validated_data) instead of using a property(validated_data) idea. And I updated its naming in my recent commits.

**Review by FarukOzderim — APPROVED** [review]
I see, LGTM 👍 , will let @aliabid94 and @abidlabs take a look as well.

File: gradio/utils.py [hunk 2]
```diff
@@ -346,3 +349,219 @@ def run_coro_in_background(func: Callable, *args, **kwargs):
...
+            exception_class(Type[Exception]): a exception type to throw with its type
+            raise_for_status(bool): a flag that determines to raise httpx.Request.raise_for_status() exceptions.
+        """
+        self._response = None
+        self._exception = None
+        self._status = None
+        self._raise_for_status = raise_for_status
+        self._validation_model = validation_model
+        self._validation_function = validation_function
+        self._exception_type = exception_type
```
Comment (FarukOzderim) [inline gradio/utils.py:420]:
```suggestion
        self._exception_type = exception_type
        self._validated_data = None
```

**Comment by FarukOzderim** [comment]
It seems that tests run forever in CircleCI, and when I run tests locally I get, digged it for some time but could not find the exact reason. After a discussion with @hibestil we think that it might due to client not being closed. He will make an update about this.
```
======================== 7 passed, 3 warnings in 4.84s ========================
Process finished with exit code 0
PASSED                              [ 14%]PASSED                             [ 28%]PASSED              [ 42%]PASSED     [ 57%]PASSED                   [ 71%]PASSED           [ 85%]Exception ignored in: <function _ProactorBasePipeTransport.__del__ at 0x0000028E2A85AD30>
PASSED  [100%]Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\proactor_events.py", line 116, in __del__
    self.close()
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\proactor_events.py", line 108, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\base_events.py", line 746, in call_soon
    self._check_closed()
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\base_events.py", line 510, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function _ProactorBasePipeTransport.__del__ at 0x0000028E2A85AD30>
Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\proactor_events.py", line 116, in __del__
    self.close()
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\proactor_events.py", line 108, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\base_events.py", line 746, in call_soon
    self._check_closed()
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\base_events.py", line 510, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function _ProactorBasePipeTransport.__del__ at 0x0000028E2A85AD30>
Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\proactor_events.py", line 116, in __del__
    self.close()
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\proactor_events.py", line 108, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\base_events.py", line 746, in call_soon
    self._check_closed()
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\base_events.py", line 510, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed

```

**Comment by FarukOzderim** [comment]
RuntimeError: Event loop is closed Is passed with the latest commit, but the tests still run forever, this is very strange indeed.

File: test/test_utils.py [hunk 3]
```diff
@@ -184,5 +186,123 @@ def test_delete_none(self):
         self.assertEqual(delete_none(input), truth)
 
 
+class TestRequest:
+    @pytest.fixture(autouse=True)
```
Comment (FarukOzderim) [inline test/test_utils.py:190]:
```suggestion
    @pytest_asyncio.fixture(autouse=True)
```

File: test/test_utils.py [hunk 4]
```diff
@@ -1,18 +1,24 @@
...
 import unittest
 import unittest.mock as mock
 import warnings
+from typing import Literal
 
 import pkg_resources
+import pytest
 import requests
+from httpx import AsyncClient
+from pydantic import BaseModel
```
Comment (FarukOzderim) [inline test/test_utils.py:14]:
```suggestion
import pytest_asyncio

```

**Comment by hibestil** [comment]
> **Found a solution:** renamed test_utils.py to 0test_utils.py and the problem resolved 🐱 Some problems occur due to other test files, changing the order of the tests fixes the problem, it might probably be related to fixtures. Shall we move fixtures to some other place @hibestil?

Thank's @FarukOzderim for your help to find the solution. I agree with you, we should create a folder for fixtures. It'll be a better way to access the same fixtures from different test files. 

I will keep the client fixture in the test file (if possible) to protect this PR's context while updating the PR with your suggestions. And we can discuss the fixture folder in another issue. Is this ok?

**Comment by FarukOzderim** [comment]
While testing&digging the situation for a while with @hibestil we found out that `test_tunneling.py` causes async tests to run forever. I also remember a similar situation when I first problematic failing tests in #572.

So for we just unblocked this PR with renaming test_tunneling to play with the order of files.
image

File: gradio/utils.py [hunk 5]
```diff
@@ -346,3 +349,220 @@ def run_coro_in_background(func: Callable, *args, **kwargs):
...
+        (1) Includes response validation functionality both using validation models and functions.
+        (2) Since we're still using httpx.Request class by wrapping it, we have all it's functionalities.
+        (3) Exceptions are handled silently during the request call, which gives us the ability to inspect each one
+        individually in the case of multiple asynchronous request calls and some of them failing.
+        (4) Provides HTTP request types with Request.Method Enum class for ease of usage
+    Request also offers some util functions such as has_exception, is_valid and status to inspect get detailed
+    information about executed request call.
+
+    The basic usage of Request is as follows: create a Request object with inputs(method, url etc.). Then use it
+    with the "await" statement, and then you can use util functions to do some post request checks depends on your case.
```
Comment (FarukOzderim) [inline gradio/utils.py:370]:
```suggestion
    with the "await" statement, and then you can use util functions to do some post request checks depending on your use-case.
```

File: gradio/utils.py [hunk 6]
```diff
@@ -346,3 +349,220 @@ def run_coro_in_background(func: Callable, *args, **kwargs):
...
+    Finally, call the get_validated_data function to get the response data.
```
Comment (FarukOzderim) [inline gradio/utils.py:371]:
```suggestion
    Finally, call the get_validated_data function to get the response data.
    
    You can see example usages in test_utils.py.
```

**Comment by FarukOzderim** [comment]
Merging this PR to unlock other depending issues, please drop your comments on the PR and we'll adress them later.

File: gradio.egg-info/requires.txt [hunk 7]
```diff
@@ -16,3 +16,5 @@ pydub
 requests
 uvicorn
 Jinja2
+httpx
```
Comment (abidlabs) [inline gradio.egg-info/requires.txt:19]: If we are going to go with `httpx`, can we replace our usage of `aiohttp` with `httpx` entirely and remove `aiohttp` as a dependency?

File: gradio/utils.py [hunk 8]
```diff
@@ -346,3 +349,222 @@ def run_coro_in_background(func: Callable, *args, **kwargs):
     """
     event_loop = asyncio.get_event_loop()
     _ = event_loop.create_task(func(*args, **kwargs))
+
+
+client = httpx.AsyncClient()
```
Comment (abidlabs) [inline gradio/utils.py:354]: Should be property of the `Request` class

File: gradio/utils.py [hunk 9]
```diff
@@ -346,3 +349,222 @@ def run_coro_in_background(func: Callable, *args, **kwargs):
...
+
+
+class Request:
```
Comment (abidlabs) [inline gradio/utils.py:357]: Should we call this class `AsyncRequest`? The name `Request` seems too generic

File: test/test_utils.py [hunk 10]
```diff
@@ -184,5 +187,124 @@ def test_delete_none(self):
...
+            url="https://reqres.in/api/users",
+            json={"name": "morpheus", "job": "leader"},
+            validation_model=TestModel,
+        )
+        assert isinstance(client_response.get_validated_data(), TestModel)
+
+    @pytest.mark.asyncio
+    async def test_validate_and_fail_with_model(self):
+        class TestModel(BaseModel):
+            name: Literal[str] = "John"
```
Comment (abidlabs) [inline test/test_utils.py:242]: Type should be "str" instead of "Literal[str]"

File: gradio/utils.py [hunk 11]
```diff
@@ -346,3 +349,222 @@ def run_coro_in_background(func: Callable, *args, **kwargs):
...
+        self._exception_type = exception_type
+        self._validated_data = None
+        # Create request
+        self._request = self._create_request(method, url, **kwargs)
+
+    def __await__(self):
+        """
+        Wrap Request's __await__ magic function to create request calls which are executed in one line.
+        """
+        return self.__run().__await__()
```
Comment (abidlabs) [inline gradio/utils.py:431]: I'm confused what's happening here. It looks like you are recursively calling the `__await__()` function?

**Comment by abidlabs** [comment]
Looks good! Left some questions / suggestions above.

File: gradio.egg-info/requires.txt (hunk 7 above)
Comment (FarukOzderim) [reply gradio.egg-info/requires.txt:19]: Yeah planning to it

File: gradio/utils.py (hunk 9 above)
Comment (FarukOzderim) [reply gradio/utils.py:357]: since it will be used as `await Request(...)` I think it is clear that it is async. Shorter names are better imo 🐱

File: gradio/utils.py (hunk 11 above)
Comment (FarukOzderim) [reply gradio/utils.py:431]: Yeah I was confused initially as well, but it was clear to me thx to @hibestil 

Request.await() ->
Runs
1. Request.init
2. Request.await
->
Runs
Request.__run

And later-on we await the async __run function.

File: test/test_utils.py (hunk 10 above)
Comment (FarukOzderim) [reply test/test_utils.py:242]: Thx for catching, I've missed that, I shall update in my current PR

File: gradio/utils.py (hunk 8 above)
Comment (abidlabs) [reply gradio/utils.py:354]: But we don't need to write `Request.client` except inside the `Request` class and perhaps for mocking tests, right? In other words, we won't need to write it out whenever we are making async requests from other parts of the `gradio` library? If that's the case, then I think it's better to put it inside the `Request` class to keep the code better organized

File: gradio/utils.py (hunk 11 above)
Comment (abidlabs) [reply gradio/utils.py:431]: Sorry, still not clear to me.

In step (2), we run  `self.__run().__await__()`, but `self.__run()` returns the Request object, so are we not running Request.__run() again and again?

File: gradio/utils.py (hunk 8 above)
Comment (FarukOzderim) [reply gradio/utils.py:354]: Yeah, I agree, will update, thanks for defending a better-organized approach 🐱

File: gradio/utils.py (hunk 11 above)
Comment (FarukOzderim) [reply gradio/utils.py:431]: Let me try again.

Awaiting a Request instance causes `__await__` function to run, which awaits the `__run` function. See [here](https://docs.python.org/3/reference/datamodel.html?highlight=__await__#awaitable-objects) for more

In Short
`await Request()`
1. creates a Request object
2. Awaits the __run function

Comment (hibestil) [reply gradio/utils.py:431]: Thanks, @FarukOzderim for your clarifications. In addition to yours, I'd like to share the process step by step:
- When an instance is created using the Request class 
 ```python
 client_response = await Request(method=Request.Method.GET, url="http://headers.jsontest.com/")
```
- `await` statement calls the` Request.__await__` function to define the behavior of the `Request` instance when it is awaited.
- In  the` Request.__await__` function, we call the `Request.__run` function to execute the request stuff. 
    - Since `__run` is an async function, we have to await it, so here we used `self.__run().__await__()` to call an async function inside a non-async function.

**Briefly:** Using the `await` statement while the creation of the Request instance leads us to execute the `Request.__run` method. So we are not calling a thing recursively.


**Reference:**
https://docs.python.org/3/reference/datamodel.html#object.__await__
